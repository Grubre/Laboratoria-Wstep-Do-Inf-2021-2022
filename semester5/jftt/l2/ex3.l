%{
#include <stdio.h>
#include <stdbool.h>

bool keep_docs = false;

int yylex(); 
%} 

ESCAPE      (\\\n.*)*

%s STRING
%s DOCS
%s MULTILINE
%s INCLUDE

%%
<INITIAL>#include[[:blank:]]*                                                                                   { ECHO; BEGIN(INCLUDE); }
<INCLUDE>(\/\*).*(\*\/)                                                                                         { }
<INCLUDE>>                                                                                                      { ECHO; BEGIN(INITIAL); }

<INITIAL>#include[[:blank:]]*\"                                                                                 { ECHO; BEGIN(STRING); }

<INITIAL>\"                                                                                                     { ECHO; BEGIN(STRING); }
<STRING>\\\\\"                                                                                                  {ECHO; BEGIN(INITIAL);}
<STRING>[^\\](\\\\)*\"                                                                                       { ECHO; BEGIN(INITIAL); }

<INITIAL>(\/{ESCAPE}\*{ESCAPE}\*)|(\/{ESCAPE}\*{ESCAPE}!)				                                        { if(keep_docs) { ECHO;} BEGIN(DOCS); };
<DOCS>{ESCAPE}\*{ESCAPE}\/						                                                                { if(keep_docs) { ECHO;} BEGIN(INITIAL); };
<DOCS>.							                                                                                { if(keep_docs) { ECHO;} };

<INITIAL>\/{ESCAPE}\*							                                                                { BEGIN(MULTILINE); };
<MULTILINE>{ESCAPE}\*{ESCAPE}\/				                                                                    { BEGIN(INITIAL); };
<MULTILINE>.                                                                                                    { }

<INITIAL>((\/{ESCAPE}\/{ESCAPE}\/)|(\/{ESCAPE}\/{ESCAPE}!)).*{ESCAPE}.*$                                        { if(keep_docs) { ECHO;} }
<INITIAL>\/{ESCAPE}\/.*{ESCAPE}.*$				                                                                { };
%%

int main(int argc, char** argv) {
    if (argc > 1 && strcmp(argv[1], "-docs") == 0) {
        keep_docs = true;
    }
    
	yylex();

	return 0;
}
