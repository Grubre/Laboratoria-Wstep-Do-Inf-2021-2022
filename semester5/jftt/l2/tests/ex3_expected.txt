#include<iostream>
#include<cstdio>










using namespace std;



int some_function(int a) {
	return a << (1 << 4);
}







int main()
{
  
  int i = 5;
  
  cout << "Jakis program" << endl; 
  
  cout << "Poczatek komentarza /*" << endl; 
  


  
  


  
  
  
  cout << "Koniec komentarza */ "<< endl; 
  cout << "Komentarz /* ala */" << endl;
  
  

  


  


  
   
  cout << "Zabawa \" // \\\\\"ala i kot \\\\" << endl; 
  cout << "Komentarz // kot " << endl;
  cout << "Pulapka \" \
           // ma \
           /* \\\"ma\" */ \
		   /** doc */\
           \\\\" << endl;
 cout << "Zabawa \" // ala i kot " << endl;	   
 printf("Zabawa \" // ala i kot ");
}
